File: python/day04/day04.py
with open('day04/input', encoding='utf-8') as f:
    pairs = [
        (
            set(range(
                int(first_elf.split('-')[0]),
                int(first_elf.split('-')[1])+1
            )),
            set(range(
                int(second_elf.split('-')[0]),
                int(second_elf.split('-')[1])+1
            ))
        )
        for first_elf, second_elf in [
            line.strip().split(',') for line in f
        ]
    ]

complete_overlap_count = len([
    '' for first_elf, second_elf in pairs
    if (
        first_elf.issubset(second_elf)
        or second_elf.issubset(first_elf)
    )
])
print(f'Part 1: {complete_overlap_count}')

overlap_count = len([
    '' for first_elf, second_elf in pairs
    if len(first_elf.intersection(second_elf)) > 0
])
print(f'Part 2: {overlap_count}')

File: typescript/src/day04/day04.ts
import fs from "fs";

const pairs = fs
  .readFileSync("src/day04/input")
  .toString("utf-8")
  .split("\n")
  .map((line) => {
    return line.split(",").map((elf) => {
      const min = parseInt(elf.split("-")[0]);
      const max = parseInt(elf.split("-")[1]);
      const size = max - min + 1;
      return [...Array(size).keys()].map((i) => i + min);
    });
  });

const totalOverlapCount = pairs
  .map((pair) => {
    return (
      pair[0].every((n) => pair[1].includes(n)) ||
      pair[1].every((n) => pair[0].includes(n))
    );
  })
  .filter((overlap) => overlap).length;
console.log("Part 1: " + totalOverlapCount);

const overlapCount = pairs
  .map((pair) => {
    return pair[0].some((n) => pair[1].includes(n));
  })
  .filter((overlap) => overlap).length;
console.log("Part 2: " + overlapCount);
